enhancements/installer/log-config.md

```diff
@@ -0,0 +1,189 @@
+---
+title: log-config
+authors:
+  - @patrickdillon
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - @rna-afk
+  - @2uasimojo
+  - @jstuever
+  - @jhixson74
+approvers:
+  - @sdodson
+  - @zaneb
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - n/a
+creation-date: 2022-07-25
+last-updated: 2022-07-25
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/CORS-2068
+see-also:
+  - https://docs.google.com/document/d/16_tccZjADKJMd2txQ7nmz8olZ5iUrw81XxGTRmqMXoU/edit#
+---
+
+# Installer Log Configuration
+
+## Summary
+
+This enhancement proposes the introduction of a log-config.yaml configuration
+file which allows users to pass input that will be used specifically for
+configuring logging behavior in the Installer.
+
+## Motivation
+
+* Decorating installer logs with k/v pairs provided by the ARO resource provider is the
+top priority ask for the installer in the 4.12 release. The ARO team wants this feature
+in order to filter logs when logs from multiple installs are pooled. For example, all installer
+logs could be decorated with a k/v pair denoting an aroClusterID:
+
+```shell
+INFO OpenShift Installer unreleased-master-6230-gf46ac2f0d1a3628a51d3437bd79e4221da9aff23-dirty  aroClusterID=123
+INFO Built from commit f46ac2f0d1a3628a51d3437bd79e4221da9aff23  aroClusterID=123
+INFO Fetching Master Machines...                   aroClusterID=123
+INFO Loading Master Machines...                    aroClusterID=123
+```
+
+* We can generalize the ask from ARO so that it may be useful to other users
+* We should keep future logging needs in mind when designing this solution 
+so that we do not complicate or preclude future development work
+
+
+### User Stories
+
+* As an installer user, I want to be able to provide a set of key-value pairs that
+will decorate all Installer log lines
+* As an installer user, I would like to configure the level of logging through a config file 
+* As an installer developer, I want the design for this user input to be extensible
+
+### Goals
+
+The goal of this enhancement is to determine the proper way of passing user input
+for the need of configuring logging.
+
+### Non-Goals
+
+The need for decorating log lines is just an example of the kind of configuration
+that is needed. The implementation for decorating log lines is not relevant
+to this discussion.
```

Comment:
The prefixing stuff feels like it should be possible to do with post-processing or a shell wrapper, rather than building anything into the installer itself. What's another example of a configuration value that the installer does need to own?

Reply:
> The prefixing stuff feels like it should be possible to do with post-processing or a shell wrapper, rather than building anything into the installer itself. 

This is a good point and it crossed my mind as well, but if this is supposed to be done for managed services, I think that responsibility would fall to hive. So I think the specific question would be does it make more sense for hive to do this postprocessing or to add support in the installer. @2uasimojo wdyt about this?

> What's another example of a configuration value that the installer does need to own?

We do not have another need that is prioritized. Another example could be if we wanted to output logs as json, but I think that could be handled without the config.

Reply:
> does it make more sense for hive to do this postprocessing

To my view, this argument could as easily be made at the higher level: why doesn't ARO just tag the logs as they grab them?

I think it boils down to simplicity of implementation [1]. The prototypes from both [hive](https://github.com/openshift/hive/pull/1828) and [installer](https://github.com/openshift/installer/pull/6161) demonstrate that this can be done in a generic and fairly robust way. Post-processing sounds more brittle, error-prone, and use case-specific to me.

For instance: it would not be sufficient to wait until the installer process is finished, in case the logs are being consumed realtime as they are produced. So we would have to pipe them through a process that decorates them on the fly. Doable, but a more disruptive change.

[1] Overall, that is. Obviously if hive/ARO does the tagging, the installer doesn't have to do anything, and that's simpler for the installer...

Reply:
I will incorporate this discussion/these perspectives into my next set of edits.

Reply:
When I said "post-processing" I wasn't being precise. A wrapper or other thing that updates the messages as they are emitted from the installer would do it. It's possible to do that with a small shell script wrapper and that could live anywhere. See https://github.com/openshift-metal3/dev-scripts/blob/master/logging.sh#L14 for one example where the script adds dynamic timestamps to all of its own stdout. Static values like key/value pairs would be even simpler.

Reply:
I understand. We would have to do the equivalent from our go code where it [execs](https://github.com/openshift/hive/pull/1828/files#diff-5e902c7f1d283bf3c230c2a091c668996e7546247fb77cb42b6c041257ff56adR976) the installer (or birth and maintain a script wrapper for the purpose). Again, not terribly difficult; it just feels hackier and more error-prone. (Also less extensible for future use cases of general log configuration, should they arise.)

Reply:
I have tried to capture the perspectives in this discussion in the most recent commits. I will leave this thread "unresolved" as it is critical